### DISCUSSION TITLE
Validating embedded address entities with different rules

### DISCUSSION THREAD

**Discussion by alcohol** [symfony/symfony#54152] [Q&A]
I am trying to figure out how to properly use the Validator component to apply certain assertions to an embedded address entity (in an order) based on whether the entity serves as a shipping address or a billing address (some rules only apply in one or the other scenario while other rules apply to both always). I thought I could do this using validation groups, but I quickly realized that is not the case (see https://github.com/symfony/symfony/issues/40741 and similar issues). I imagine this is not a very uncommon scenario,  so I was wondering if there is a recommended approach to handle this? I could not find anything in the documentation that really nudges me in the right direction. To make things a bit more complicated, the user can specify that the shipping address is identical to the billing address, in which case both scenarios apply to the same embedded entity.

```php
class Order {
  private Address $billing;
  private ?Address $shipping;
}
```

**Comment by MatTheCat** [comment]
Maybe I’m missing something here but I would just implement specific constraints/validators for specific contexts, like a `BillingAddress`/`Validator` e.g.

You’d still use `Valid` to ensure the address is valid, and maybe wrap the two constraints into a [`Sequentially`](https://symfony.com/doc/current/reference/constraints/Sequentially.html) one.

```php
use App\Validator\Constraints\{BillingAddress, ShippingAddress}
use Symfony\Component\Validator\Constraints as Assert;

class Order {
  #[Assert\Sequentially([
    new Assert\Valid,
    new BillingAddress,
  ])]
  private Address $billing;

  #[Assert\Sequentially([
    new Assert\Valid,
    new ShippingAddress,
  ])]
  private ?Address $shipping;
}
```

That way `Address` does not need to be aware of every context it can be used in.

Would that work for you?

**Reply by alcohol** [reply]
Not only is there a difference between them when it comes to billing or shipping context though, but based on country the validation rules for postalcodes or other things vary as well. And to make things worse, when the customer checks the box "ship to billing address", there is no shipping address entity value and all validation should be applied to the billing address entity. So I'm starting to think attributes are not the way to go, but rather dynamically through form configuration (using the form events to determine where to apply certain conditional validations).

**Reply by MatTheCat** [reply]
> based on country the validation rules for postalcodes or other things vary as well

Sounds like the responsibility of an `AddressValidator`; does this cause any issue?

> when the customer checks the box "ship to billing address", there is no shipping address entity value and all validation should be applied to the billing address entity

IMO "ship to billing address" is better represented by the shipping address being the same than the billing’s rather than no shipping address. I understand you want to map your form model to the UI, but seeing https://github.com/symfony/symfony/discussions/54152#discussioncomment-8690580 it may be simpler and clearer to have the form populate `Order::$shipping` if the checkbox if checked (cannot say for sure since I did not try it though ¯\_(ツ)_/¯).

**Reply by alcohol** [reply]
`AddressValidator` is not an issue immediately. But I think it does have the tendency to become a place to dump anything "validate street/postalcode/name/etc" related, which can quickly turn it into a very large class with a lot of responsibilities.

Using the same `Address` entity for both properties is a possibility. I would need to think about how that works out templating wise though. As soon as you untick the box "ship to billing address", the shipping address fields would show up prefilled with the values of the entity that backs the billing address form as well. When you edit those values though, one would have to make sure a new entity is created first rather than the same entity (billing) being updated.

**Comment by alcohol** [comment]
Using the following approach right now; 
- update billing field to also apply shipping validation if same address toggle is used,
- update shipping field to apply shipping validation if same address toggle is not used,
- even when not used, we keep the shipping field because we still render it in html (hidden).

```php
$builder
    ->add(
        'billingAddress',
        OrderAddressType::class,
        [
            'countries' => $countries,
            'required' => true,
            'validation_groups' => ['Default', 'billing'],
        ]
    )
    ->add(
        'shippingAddress',
        OrderAddressType::class,
        [
            'countries' => $countries,
            'required' => false,
        ]
    )
;

/**
 * Update billing or shipping field based on submitted data.
 *
 * @see https://symfony.com/doc/current/form/events.html#2-submitting-a-form-formevents-pre-submit-formevents-submit-and-formevents-post-submit
 */
$builder->addEventListener(
    FormEvents::PRE_SUBMIT,
    static function (FormEvent $event) use ($countries): void {
        /** @var mixed[] $data */
        $data = $event->getData();
        $form = $event->getForm();

        if (CustomerDetails::DELIVERY_TYPE_ALTERNATIVE === $data['deliveryType']) {
            // redefine field
            $form->remove('shippingAddress')->add(
                'shippingAddress',
                OrderAddressType::class,
                [
                    'countries' => $countries,
                    'required' => true, // mark as required
                    'validation_groups' => ['Default', 'shipping'], // add validation groups
                    'constraints' => [new Valid(groups: ['Default', 'shipping'])], // add Valid constraint
                ]
            );
        } else {
            // redefine field
            $form->remove('billingAddress')->add(
                'billingAddress',
                OrderAddressType::class,
                [
                    'countries' => $countries,
                    'required' => true,
                    'validation_groups' => ['Default', 'billing', 'shipping'], // also apply shipping validation
                    'constraints' => [new Valid(groups: ['Default', 'billing', 'shipping'])], // override default Valid constraint
                ]
            );

            // redefine field
            $form->remove('shippingAddress')->add(
                'shippingAddress',
                OrderAddressType::class,
                [
                    'countries' => $countries,
                    'required' => false, // mark as not required
                    'mapped' => false, // do not map to model
                    'validation_groups' => false, // do not validate
                ]
            );

            // discard submitted data
            unset($data['shippingAddress']);

            $event->setData($data);
        }
    }
);
```
